kinto/core/resource/__init__.py

```diff
@@ -460,17 +464,21 @@ def patch(self):
         existing = self._get_record_or_404(self.record_id)
         self._raise_412_if_modified(existing)
 
-        try:
-            # `data` attribute may not be present if only perms are patched.
-            changes = self.request.json.get('data', {})
-        except ValueError:
-            # If no `data` nor `permissions` is provided in patch, reject!
-            # XXX: This should happen in schema instead (c.f. ShareableViewSet)
-            error_details = {
-                'name': 'data',
-                'description': 'Provide at least one of data or permissions',
-            }
-            raise_invalid(self.request, **error_details)
+        # patch is specified as a list of of operations (RFC 6902)
+        if self._is_json_patch:
+            changes = {'data': self.request.json}
```

Comment:
Yes. That's correct, and it may be a strange approach, but was the way that I've found to avoid modifying ``apply_changes`` signature, without patching arguments early on update, or avoiding updating changes after the ``apply_changes`` call. Perhaps we could use a keyword argument to pass something like ``operations`` and ``changes = {}`` to ``apply_changes``. It would still be a bidirectional argument, but never on the same call.